File: G00000001 - Dominic Carr - Assignment 1/python/shop.py
# shop.py
# A simple shopping simulation in procedural Python
# Author: Fiachra O' Donoghue

# 1. stock.csv holds stock and wholesale prices
# 2. Customer prices are MARKUP greater than wholesale
# 3. MARKUP doesn't change during a programme run
# 4. Initial cash shop holds is in stock.csv
# 5. Modes: Fully automatic: Customers generated on 
#               the fly with random amount of money, random shopping list, 
#               when shop runs out of something it spends 50% of its
#               remaining money on that thing. Simulation runs until shop
#               can't restock.
#           Fully Manual: There are ten pregenerated cusomers in the customer
#                         list. They all shop on by one.
#           The user is the customer. Tell the programme your name and how much 
#           you have and then start buying things.



from dataclasses import dataclass, field
from typing import List
import csv
import random
import time
from faces import faces

# Constants
# Paths to stock and names files
STOCK_PATH = '../stock.csv'
NAMES_PATH = '../names.list'
CUSTOMERS_PATH = '../customers.csv'

# Value ranges for random customer generation
BUDGET_RANGE = (100, 400)
ITEMS_RANGE = (1, 5)
PIECES_RANGE = (1, 100)
REORDER_THRESHOLD = 5

shopkeeper = r"""
       _www_ 
      /-o-o-\  
    (|   -   |)
      \ -=- /  
      /`---'\
"""

def clear_console():
    print("\n" * 100)

# Each product has a name and a price
@dataclass
class Product:
    name: str
    price: float
    max_quantity: int

# A product's quantity
@dataclass 
class ProductStock:
    product: Product
    quantity: int

# The shop has an amount of cash and a list of products
@dataclass 
class Shop:
    cash: float = 0.0
    stock: List[ProductStock] = field(default_factory=list)

# A customer has a name, a budget, 
# a shopping list --> list of Products and quantity desired
# a receipt --> list of Products and quantity purchased
@dataclass
class Customer:
    name: str = ""
    face: str = ""
    budget: float = 0.0
    shopping_list: List[ProductStock] = field(default_factory=list)
    receipt: List[ProductStock] = field(default_factory=list)


def generate_shop():
    """Load stock from CSV file, populate Product and ProductStock data 
       structures, place in shop's stock list, and return populated shop.

    Returns:
        Shop: The fully initialised shop.
    """

    # Initialise a Shop dataclass
    shop = Shop()
    # Read csv using location stored in STOCK_PATH constant
    with open(STOCK_PATH, 'r') as f:
        csv_reader = csv.reader(f)
        # The first row holds the shop's cash
        shop.cash = float(next(csv_reader)[0])
        # Construct a list of lists. 
        # Each list contains a product name, price, and quantity
        items = [row for row in csv_reader]

    # Create Product and ProductStock data structures and 
    # add to the shop's stock field
    for item in items:
        # Create a Product dataclass (name, price, max_quantity)
        product = Product(item[0], float(item[1]), int(item[2]))
        quantity = int(item[2])
        stock = ProductStock(product, quantity)
        shop.stock.append(stock)

    return shop


def generate_customer(shop, budget_range, names_path, items_range, pieces_range):
    """Generate a random customer for the shop. The customer will have a random name, 
    a budget randomly chosen from within a specified range, and a shopping list with a 
    random number of items and a random quantity of each item. The number of items and 

    Args:
        shop ([type]): [description]
        budget_range ([type]): [description]
        items_range ([type]): [description]
        pieces_range ([type]): [description]

    Returns:
        [type]: [description]
    """
    customer = Customer()
    with open(names_path, 'r') as f:
        names = f.readlines()
        names = [name.rstrip() for name in names]
        customer.name = random.choice(names)

    customer.face = random.choice(faces)

    customer.budget = random.randint(budget_range[0], budget_range[1])
    num_items = random.randint(items_range[0], items_range[1])
    items = random.sample(shop.stock, num_items)

    for item in items:
        num_pieces = random.randint(pieces_range[0], pieces_range[1])
        customer.shopping_list.append(ProductStock(item.product, num_pieces))

    return customer

#def load_customers():

def stringify_product(product):
    return f"{product.name:<15}{product.price:>7.2f}"

def stringify_shop(shop):
    stock = line = "-" * 36 + "\n"
    stock += ("    Item" + " " * 13 + "Price" + " " * 5 + "Stock\n" + line)
    for i, item in enumerate(shop.stock):
        stock += (f"{i + 1:>2}. " + stringify_product(item.product) + 
                  f"{item.quantity:>10}" + "\n")
    stock += line
    cash = f"Cash: {shop.cash:>7.2f}"
    return stock + "\n" + cash + "\n" + line

def total_bill(itemlist):
    total = 0
    for item in itemlist:
        total += item.product.price * item.quantity
    return total

def stringify_bill(itemlist):
    out = line = "-" * 49 + "\n"
    out += ("    Item" + " " * 13 + "Price" + " " * 5 + "Quantity" + " " * 5 + "Total\n" + line)
    for i, item in enumerate(itemlist):
        out += (f"{i + 1:>2}. {stringify_product(item.product)}" + 
                          f"{item.quantity:>13}" +
                          f"{item.product.price * item.quantity:>10.2f}\n")
    out += line
    total = f"Total Cost: {total_bill(itemlist):37.2f}"

    return out + total + "\n" + line

def stringify_customer(customer):
    budget = f"Budget: {customer.budget:>7.2f}"
    return f"{customer.face}\nName: {customer.name}\n{budget}\n{stringify_bill(customer.shopping_list)}"

def stringify_receipt(customer):
    return stringify_bill(customer.receipt)

def transact(shop, customer):
    for item in customer.shopping_list:
        for stock in shop.stock:
            if item.product == stock.product:
                # If customer wants more than shop has
                if item.quantity > stock.quantity:
                    print(f"{customer.name} wants {item.quantity} x {item.product.name} but only {stock.quantity} are available.")
                    item.quantity = stock.quantity
                if customer.budget < item.product.price * item.quantity:
                    item.quantity = int(customer.budget / item.product.price)
                    if item.quantity == 0:
                        print(f"{customer.name} can't afford any {item.product.name}.")
                    else:
                        print(f"{customer.name} can only afford {item.quantity} x {item.product.name}")
                print(f"{customer.name} buys {item.quantity} x {item.product.name} for {item.product.price * item.quantity:.2f}")
                stock.quantity -= item.quantity
                shop.cash += item.product.price * item.quantity
                customer.budget -= item.product.price * item.quantity
                customer.receipt.append(item)
                print(f"{customer.name} has \N{euro sign}{customer.budget:.2f} left.\n")

    shop.cash += total_bill(customer.shopping_list)
    customer.budget -= total_bill(customer.shopping_list)
    input("Press Enter to continue.")
    clear_console()
    print(shopkeeper)
    print(f"Thank you {customer.name}. Here's your receipt. \nPlease come again.\n")
    print(f"{customer.name}'s Receipt:")
    print(stringify_bill(customer.receipt))


def restock(shop, product):
    for item in shop.stock:
        if item.product == product:
            current_quantity = item.quantity
            item.quantity = item.product.max_quantity
            shop.cash -= item.product.price * (item.product.max_quantity - current_quantity)


def check_stock(shop, reorder_threshold):
    out_of_stock = []
    for item in shop.stock:
        if item.quantity <= reorder_threshold:
            out_of_stock.append(item.product.name)
            restock(shop, item.product)
    return out_of_stock


def auto_mode():
    shop = generate_shop()
    while True:
        clear_console()
        customer = generate_customer(shop, BUDGET_RANGE, NAMES_PATH, ITEMS_RANGE, PIECES_RANGE)
        print(f"{customer.name} has come into the shop with a shopping list!")
        print(stringify_customer(customer))
        input("Press enter to begin transaction.")
        clear_console()
        print(f"{customer.face}")
        print(f"{customer.name} is shopping.\n")
        transact(shop, customer)
        input("Press Enter to continue...")

        # Check if any items are out of stock
        clear_console()
        print(shopkeeper)
        print("Better take a look in the stockroom.")
        out_of_stock = check_stock(shop, REORDER_THRESHOLD)
        if len(out_of_stock) == 0:
            print("Everything seems ok.")
        else:
            if len(out_of_stock) > 1:
                out_of_stock[-1] = "and " + out_of_stock[-1]
            print(f"Yikes, I'm almost out of {out_of_stock[0]} {', '.join(out_of_stock[1:])}")
            print("Better reorder")
        print(stringify_shop(shop))
        input("Press Enter to continue...")

#def preset_mode(customers_path, shop):
    


def live_mode():
    pass

def load_state():
    pass

def save_state():
    pass

def generate_customers(num_customers, shop, budget_range, names_path, items_range, pieces_range, file_path):
    print("Generating customers...")
    with open(file_path, 'w') as f:
        writer = csv.writer(f)
        for i in range(num_customers):
            customer = generate_customer(shop, budget_range, names_path, items_range, pieces_range)
            writer.writerow([customer.name, f"{float(customer.budget):.2f}"])
            for item in customer.shopping_list:
                writer.writerow([item.product.name, item.quantity])
            writer.writerow(['----------'])
    print(f"Generated {num_customers} customers.")
    input("Press Enter to continue.")

def print_help():
    pass

# Generate menu
def displayMenu():
    clear_console()
    line = "-" * 19
    menuItems = ["Auto Mode",
                 "Preset Mode",
                 "Live Mode",
                 "Load State",
                 "Save State",
                 "Generate Customers"]
    
    print(line)
    print("Shop Simulator 2000")
    print(shopkeeper)
    print("Hi I'm the shopkeeper.\nHow can I help?")
    print(line)
    print("MENU")
    print("----")

    for index, item in enumerate(menuItems):
        print(index + 1, "-", item)

    print("x - Exit Application")


def main():

    shop = generate_shop()

    while True:

        displayMenu()

        selection = input("Choice: ")

        if selection == "1":
            auto_mode()
        elif selection == "2":
            preset_mode()
        elif selection == "3":
            live_mode()
        elif selection == "4":
            load_state()
        elif selection == "5":
            save_state()
        elif selection == "6":
            generate_customers(10, shop, BUDGET_RANGE, NAMES_PATH, ITEMS_RANGE, PIECES_RANGE, CUSTOMERS_PATH)
        elif selection == "7":
            print_help()
        elif selection == "x":
            break
        else:
            print("\nPlease enter a number between 1 and 6, or x to quit")
            continue

if __name__ == "__main__":
    main()
    # shop = generate_shop()
    # customer = generate_customer(shop, BUDGET_RANGE, NAMES_PATH, ITEMS_RANGE, PIECES_RANGE)
    # print(stringify_customer(customer))
    # transact(shop, customer)
 

File: G00398776- Fiachra O Donoghue - Assignment 1/java/shop/Shop.java
package shop;

import java.io.IOException;
import java.io.PrintWriter;
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.charset.StandardCharsets;
import java.util.Random;
import java.util.Scanner;
import java.io.PrintWriter;

public class Shop {

    private double cash;
    private List<ProductStock> stock;

    public Shop(String fileName) {

        stock = new ArrayList<>();
        List<String> lines = Collections.emptyList();
		try {
			lines = Files.readAllLines(Paths.get(fileName), StandardCharsets.UTF_8);
			cash = Double.parseDouble(lines.get(0));
			// i am removing at index 0 as it is the only one treated differently
			lines.remove(0);
			for (String line : lines) {
				String[] arr = line.split(",");
				String name = arr[0];
				double price = Double.parseDouble(arr[1]);
				int quantity = Integer.parseInt(arr[2].trim());
				Product p = new Product(name, price, quantity);
				ProductStock s = new ProductStock(p, quantity);
				stock.add(s);
			}
		}

		catch (IOException e) {
			System.out.println("Cannot locate Stock file");
			e.printStackTrace();
			System.exit(-1);
		}
    }

    private List<Customer> loadCustomers(String fileName) {
        String fullPath = System.getProperty("user.dir") + fileName;
        List<Customer> customers = new ArrayList<>();
        List<String> lines = Collections.emptyList();
        try {
            boolean id = true;
            lines = Files.readAllLines(Paths.get(fullPath), StandardCharsets.UTF_8);
            for (String line : lines) {
                // End of a record
                if (line.matches("-+")) {
                    id = true;
                    continue;
                }
                // End of the file
                if (line.matches("\s*")) {
                    break;
                }
                // Start of a record
                if (id) {
                    String[] idRec = line.split(",", 2);
                    customers.add(new Customer(idRec[0], Double.parseDouble(idRec[1])));
                    id = false;
                }
                // Shopping list line
                else {
                    String[] arr = line.split(",");
                    String name = arr[0];
                    int quantity = Integer.parseInt(arr[1].trim());
                    // Find the product in the stock
                    for (ProductStock ps : stock) {
                        if (ps.getProduct().getName().equalsIgnoreCase(name)) {
                             customers.get(customers.size() - 1).addToShoppingList(new ProductStock(ps.getProduct(), quantity));
                        }
                    }
                }
            }
        }
        catch (IOException e) {
            System.out.println("Cannot locate Customers file");
            e.printStackTrace();
            System.exit(-1);
        }
        return customers;
    }      

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(Configuration.LINE_MED + System.lineSeparator());
        sb.append("  Item             Price        Stock" + System.lineSeparator());
        sb.append(Configuration.LINE_MED + System.lineSeparator());
        int count = 1;
        for (ProductStock ps : stock) {
            sb.append(count + " " + ps.toString() + System.lineSeparator());
            count++;
        }
        sb.append(Configuration.LINE_MED + System.lineSeparator());
        sb.append(System.lineSeparator());
        sb.append(String.format("Shop's Cash: %17s%.2f%n", "€", cash));
        sb.append(Configuration.LINE_MED + System.lineSeparator());
        return sb.toString();
    }

    public double getCash() {
        return this.cash;
    }

    public List<ProductStock> getStock() {
        return stock;
    }

    private List<Product> checkStock(int reorder_threshold) {
        List<Product> outOfStock = new ArrayList<>();
        for (ProductStock ps : stock) {
            if (ps.getQuantity() < reorder_threshold) {
                outOfStock.add(ps.getProduct());
            }
        }
        return outOfStock;
    }

    private void restock(Product p) {
        for (ProductStock ps : stock) {
            if (ps.getProduct().equals(p)) {
                int currentQuantity = ps.getQuantity();
                ps.setQuantity(p.getMaxQuantity());
                // Shop pays for restocking
                this.cash -= p.getPrice() * (p.getMaxQuantity() - currentQuantity);
            }
        }
    }

    // Shop is responsible for parts of customer that are dependent on it
    // Customer provides its own name and face
    private Customer randomCustomer() {
        Random random = new Random(System.currentTimeMillis());
        double budget = random.nextDouble() * (Configuration.BUDGET_MAX - Configuration.BUDGET_MIN) + Configuration.BUDGET_MIN;
        int numItems = random.nextInt(Configuration.ITEMS_MIN, Configuration.ITEMS_MAX + 1);
        List<ProductStock> shoppingList = new ArrayList<>();
        for (int i = 0; i < numItems; i++) {
            int numPieces = random.nextInt(Configuration.PIECES_MIN, Configuration.PIECES_MAX + 1);
            List<ProductStock> stock = this.getStock();
            Product item = stock.get(random.nextInt(stock.size())).getProduct();
            shoppingList.add(new ProductStock(item, numPieces));
        }
        Customer customer = new Customer(budget, shoppingList);
        return customer;
    }

    public void doAutoMode() {
        while (true) {
            Customer customer = randomCustomer();
            this.shopVisit(customer);
        }
    }

    public void doPresetMode() {
        double totalTake = 0;
        List<Customer> customers = loadCustomers(Configuration.CUSTOMERS_PATH);
        for (Customer customer : customers) {
            totalTake += this.shopVisit(customer);
        }
        Utility.clearConsole();
        System.out.println(ShopSim.shopkeeper);
        System.out.println("Whew, that was a long day!");
        if (totalTake == 0) {
            System.out.println("And I only broke even!!!");
        }
        else if (totalTake > 0) {
            System.out.println(String.format("At least I made %s %.2f %s%n", "€", totalTake, "profit."));
        }
        else {
            System.out.println(String.format("%s %s%.2f\n\n", "And after all that I lost", "€", Math.abs(totalTake)));
        }
        if (Utility.contOrQuit()) {
            ShopSim.menuControl();
        }
        ShopSim.menuControl();
    }

    /* Live Mode */
    public void doLiveMode(boolean seen) {

        // Get price of cheapest item
        double minPrice = 0;
        for (ProductStock ps : stock) {
            if (minPrice == 0 || ps.getProduct().getPrice() < minPrice) {
                minPrice = ps.getProduct().getPrice();
            }
        }

        // Get array of lower-case product names
        List<String> names = new ArrayList<>();
        for (ProductStock ps : stock) {
            names.add(ps.getProduct().getName().toLowerCase());
        }

        // Clear console and show shopkeeper
        Utility.clearConsole();
        System.out.println(ShopSim.shopkeeper);

        // Greet Customer and elicit budget
        if (seen) {
            System.out.println("Oh hello. It's you again.");
		    System.out.println("How much money have you got today?");
        }
        else {
            System.out.println("Hello, I'm the shopkeeper. Welcome to my shop.");
            System.out.println("I hope you don't mind me asking but how much money have you got?");
            System.out.print("€ ");
		    seen = true;
        }
        double budget;
        Scanner scanner = new Scanner(System.in);
        while (true) {
            try {
                budget = Double.parseDouble(scanner.nextLine());
                if (budget < 0) {
                    throw new NumberFormatException();
                }
                System.out.println("Well it's not much but hopefully you'll be able to find something you can afford.");
                break;
            }
            catch (NumberFormatException e) {
                System.out.println("Please enter a positive number.");
            }
        }

        // The Live Mode loop
        while (true) {
            // Make sure the customer has enough money to buy something
            if (budget < minPrice) {
                if (budget == 0) {
                    System.out.println("You don't have any money left.");
                }
                else {
                    System.out.println(String.format("You've only got €%.2f left", budget));
                    System.out.println(String.format("The cheapest item I have costs €$,2f", minPrice));
                    System.out.println("Come back when you have more money.");
                    System.out.println("Press Enter to continue.");
                    Utility.waitForInput();
                    ShopSim.menuControl();
                }
            }

            // Customer is presented with a list of products
            System.out.println("Would you like to see my stock (y/n)?");
            scanner = new Scanner(System.in);

            String input = scanner.nextLine();
            if (input.equalsIgnoreCase("Y")) {
                Utility.clearConsole();
                System.out.println(this);
                System.out.println(ShopSim.shopkeeper);
                System.out.println("You can select a product to buy by number or name.");
            }
            else {
                System.out.println("OK, Looks like you know what you want.");
                System.out.println("So what'll it be?");
            }

            // Customer selects product
            Product selectedProduct = null;
            System.out.print("Enter product number or name (or 'Q' to end transaction): ");
            scanner = new Scanner(System.in);

            // If a number has been entered...
            if (scanner.hasNextInt()){
                int productNum = scanner.nextInt();
                if (productNum < 0 || productNum >= this.getStock().size() + 1 ) {
                    System.out.println("That's not a valid selection.");
                    continue;
                }
                else {
                    // The products are numbered from 1 but indexed from 0
                    productNum--;
                    selectedProduct = this.stock.get(productNum).getProduct();                    
                }
            }

            // If a String has been entered...
            else if (scanner.hasNextLine()) {
                input = scanner.nextLine();
                // If its a 'Q then quit'
                if (input.equalsIgnoreCase("Q")) {
                    Utility.clearConsole();
                    System.out.println(ShopSim.shopkeeper);
                    System.out.println("Thanks for shopping! Come back soon!");
                    System.out.println("Press enter to continue");
                    Utility.waitForInput();
                    ShopSim.menuControl();
                }

                // If the string is a valid product name...
                else if (names.contains(input.toLowerCase())) {
                    int index = names.indexOf(input.toLowerCase());
                    selectedProduct = this.stock.get(index).getProduct();
                }
                else {
                    System.out.println("We're all out of that.");
                    Random random = new Random(System.currentTimeMillis());
                    int randomProduct = random.nextInt(this.getStock().size());
                    System.out.println(String.format("How about some nice %s instead?", 
                        this.getStock().get(randomProduct).getProduct().getName()));
                    continue;
                }
            }
            else {
                System.out.println("That's not a valid selection.");
                continue;
            }
            // Get desired quantity
            System.out.println(String.format("%s costs €%.2f", selectedProduct.getName(), selectedProduct.getPrice()));
            System.out.println("How many would you like?");
            int quantity;
            while (true) {
                try {
                    quantity = Integer.parseInt(scanner.nextLine());
                    if (quantity < 0) {
                        throw new NumberFormatException();
                    }
                    break;
                }
                catch (NumberFormatException e) {
                    System.out.println("Please enter a positive number.");
                }
            }

            // Get available quantity of requested product
            int quantityAvailable = 0;
            ProductStock selectedProductStock = null;
            for (ProductStock ps : this.stock) {
                if (ps.getProduct().equals(selectedProduct)) {
                    quantityAvailable = ps.getQuantity();
                    selectedProductStock = ps;
                }
            }

            // if customer wants more than is in stock
            if (quantity > quantityAvailable) {
                System.out.println(String.format("Sorry, I don't have that many %ss.\n", selectedProduct.getName()));
                // the customer's quantity is adjusted to match remaining stock
                quantity = quantityAvailable;
            }
            // if the customer wants more than they can afford
            if (quantity * selectedProduct.getPrice() > budget) {
                // the customer's quantity is adjusted to match remaining budget
                quantity = (int) (budget / selectedProduct.getPrice());
                if (quantity == 0) {
                    System.out.println(String.format("You can't afford any %s.", selectedProduct.getName()));
                    continue;
                }
                else {
                    System.out.println("You can't afford that many.");
				    System.out.println(String.format("You can only get %d with €%.2f", quantity, budget));
                }
            }
            // Make an offer and elicit confirmation
            System.out.println(String.format("I can give you %d for €%.2f", quantity, quantity * selectedProduct.getPrice()));
            System.out.println("Is that ok (y/n)?");

            input = scanner.nextLine();
            if (input.equalsIgnoreCase("Y")) {
                // Customer pays
                budget -= quantity * selectedProduct.getPrice();
                // Stock is adjusted
                selectedProductStock.setQuantity(quantityAvailable - quantity);
                // Shop gets paid
                this.cash += quantity * selectedProduct.getPrice();
            }
            else {
                System.out.println("OK, Never mind.");
            }
            // Customer can leave or continue shopping
            System.out.println(String.format("You have €%.2f left.", budget));
            System.out.println("Would you like to buy something else (y/n)?");
            input = scanner.nextLine();
            if (input.equalsIgnoreCase("Y")) {
                continue;
            }
            else {
                Utility.clearConsole();
                System.out.println(ShopSim.shopkeeper);
                System.out.println("OK. Thanks for shopping! Come back soon!");
                System.out.println("Press Enter to continue");
                Utility.waitForInput();
                ShopSim.menuControl();
            }
        }
    } 

    public void generateCustomers(int numCustomers) {
        String customersPath = System.getProperty("user.dir") + Configuration.CUSTOMERS_PATH;
        try {
            PrintWriter writer = new PrintWriter(customersPath, "UTF-8");
            for (int i = 0; i < numCustomers; i++) {
                Customer customer = randomCustomer();
                writer.println(String.format("%s, %.2f", customer.getName(), customer.getBudget()));
                for (ProductStock ps : customer.getShoppingList()) {
                    writer.println(String.format("%s, %d", ps.getProduct().getName(), ps.getQuantity()));
                }
                writer.println("----------");
            }
        }
        catch (IOException e) {
            System.out.println("Error writing to file");
        }

    }
        
    

    public void transact(Customer customer) {
        // Search for the item the customer is looking for in the shop's stock
        for (ProductStock ps : stock) {
            for (ProductStock itemstock : customer.getShoppingList()) {
                // If it's found...
                if (ps.getProduct().equals(itemstock.getProduct())) {
                    // If the customer wants more than the shop can supply...
                    if (itemstock.getQuantity() > ps.getQuantity()) {
                        System.out.println(String.format("%s wants %d x %s but only %d are available.", 
                            customer.getName(), itemstock.getQuantity(), itemstock.getProduct().getName(), ps.getQuantity()));
                            // The customer reduces the quantity they want to buy to what they can get
                            itemstock.setQuantity(ps.getQuantity());
                    }
                    // If the customer can' afford all they want, the amount they can afford is calculated
                    if (customer.getBudget() < itemstock.getCost()) {
                        itemstock.setQuantity((int)(customer.getBudget() / itemstock.getUnitPrice()));

                        if (itemstock.getQuantity() == 0) {
                            System.out.println(String.format("%s can't afford any %s.", customer.getName(), itemstock.getProduct().getName()));
                        }
                        else {
                            System.out.println(String.format("%s can only afford %d x %s.", customer.getName(), itemstock.getQuantity(), itemstock.getProduct().getName()));
                        }

                    }
                    // The transaction...
                    System.out.println(String.format("%s buys %d x %s for %.2f.", 
                        customer.getName(), itemstock.getQuantity(), itemstock.getProduct().getName(), itemstock.getCost()));
                    // Deduct stock from shop
                    ps.setQuantity(ps.getQuantity() - itemstock.getQuantity());
                    // Deduct money from customer
                    customer.setBudget(customer.getBudget() - itemstock.getCost());
                    // Add money to shop
                    cash += itemstock.getCost();
                    // Add purchase to customer's receipt
                    customer.addToReceipt(itemstock);
                    System.out.println(String.format("%s has  €%.2f left.%n", customer.getName(), customer.getBudget()));
                }
            }
        }
        if (Utility.contOrQuit()) {
            ShopSim.menuControl();
        }
        Utility.clearConsole();
        System.out.println(ShopSim.shopkeeper);
        System.out.println(String.format("Thank you %s. Here's your receipt. %nPlease come again.", customer.getName()));
        System.out.println(customer.toStringWithTotal(customer.getReceipt()));
    }

    public double shopVisit(Customer customer) {
        double cacheCash = this.getCash();
        Utility.clearConsole();
        System.out.println(String.format("%s has come into the shop with a shopping list!%n", customer.getName()));
        System.out.println(customer);
        if (Utility.contOrQuit()) {
            ShopSim.menuControl();
        }

        Utility.clearConsole();
        System.out.println(customer.getFace());
        System.out.println(String.format("%s is shopping", customer.getName()));

        transact(customer);

        if (Utility.contOrQuit()) {
            return 0;
        }

        // Check stock levels
        Utility.clearConsole();
        System.out.println(ShopSim.shopkeeper);
        System.out.println(String.format("Better take a look in the stockroom.%n%n"));
        List<Product> outOfStock = this.checkStock(Configuration.REORDER_THRESHOLD);
        if (Utility.contOrQuit()) {
            ShopSim.menuControl();
        }
        Utility.clearConsole();
        System.out.println(ShopSim.shopkeeper);
        if (outOfStock.size() > 0) {
            System.out.println("The following items are out of stock:");
            for (int i = 0; i < outOfStock.size(); i++) {
                System.out.println(String.format("%s%n", outOfStock.get(i).getName()));
            }
            System.out.println(this);
            if (Utility.contOrQuit()) {
                ShopSim.menuControl();
            }
            Utility.clearConsole();
            System.out.println(ShopSim.shopkeeper);
            for (Product p : outOfStock) {
                System.out.println(String.format("Restocking %s...%n", p.getName()));
                restock(p);
            }
        }
        else {
            System.out.println(String.format("Everything seems to be ok.%n"));
        }

        System.out.println(this);
        if (Utility.contOrQuit()) {
            ShopSim.menuControl();
        }
        Utility.clearConsole();

        double take = this.getCash() - cacheCash;

        return take;

    }
    
}